src/kernels/webgl/reverse_packed_gpu.ts

```diff
@@ -0,0 +1,80 @@
+/**
+ * @license
+ * Copyright 2019 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =============================================================================
+ */
+
+import {getChannels} from '../packing_util';
+import {GPGPUProgram} from './gpgpu_math';
+import {getCoordsDataType} from './shader_compiler';
+
+export class ReversePackedProgram implements GPGPUProgram {
+  variableNames = ['x'];
+  outputShape: number[];
+  userCode: string;
+  usesPackedTextures = true;
+
+  constructor(xShape: number[], axis: number[]) {
+    const rank = xShape.length;
+    if (rank > 4) {
+      throw new Error(
+          `WebGL backend: Reverse of rank-${rank} tensor is not yet supported`);
+    }
+    this.outputShape = xShape;
+    const channels = getChannels('rc', rank);
+    const nextColumn =
+        `++${channels[rank - 1]} < ${this.outputShape[rank - 1]}`;
```

Comment:
I realize that we have this pattern elsewhere in the packed shaders, but could we instead test for adding 1 to the coordinate rather than mutating the coordinate? That way we don't have to remember to clean up after ourselves and decrement the coordinate. I think it'll be a little bit easier to read as well.

Reply:
Thanks @annxingyuan, I refactored the code a bit and updated the compiled shader code at the top of the PR as well.  Let me know what you think.